engine/src/main/java/org/camunda/bpm/engine/impl/persistence/entity/VariableInstanceEntity.java

```diff
@@ -278,26 +275,52 @@ public TypedValue getTypedValue(boolean deserializeValue) {
     return typedValueField.getTypedValue(deserializeValue, isTransient);
   }
 
-  public void setValue(TypedValue value) {
-    // clear value fields
-    clearValueFields();
+  public void setValue(TypedValue newValue) {
+    clearValueFieldsBasedOnType(newValue);
 
-    typedValueField.setValue(value);
+    typedValueField.setValue(newValue);
+  }
+
+  public void clearValueFieldsBasedOnType(TypedValue newValue) {
+    if (variableWillChangeType(newValue)) {
+      clearValueFieldsBasedOnType(true);
+    } else {
+      clearValueFieldsBasedOnType(false);
+    }
   }
 
-  public void clearValueFields() {
+  public void clearValueFieldsBasedOnType(boolean deleteVariable) {
     this.longValue = null;
     this.doubleValue = null;
     this.textValue = null;
     this.textValue2 = null;
     typedValueField.clear();
 
-    if(byteArrayField.getByteArrayId() != null) {
+    if (deleteVariable && byteArrayField.getByteArrayId() != null) {
       deleteByteArrayValue();
       setByteArrayValueId(null);
     }
   }
 
+  public boolean variableWillChangeType(TypedValue newValue) {
+    String currentTypeName = typedValueField.getTypeName();
+    String newTypeName = getTypeName(newValue);
+
+    return !Objects.equals(currentTypeName, newTypeName);
```

Comment:
❓ Can't we do simple string compare instead of object as they are both strings?

Reply:
It was selected to make the code null-proof. But i get if the types are null, the invocation code is not null-proof either.

Using simple string comparison as suggested.